### PR TITLE
Alerting: moving data source uid to query instead of model

### PR DESCRIPTION
**Pull request by mckn**
**What this PR does / why we need it**:
I have changed the QueryEditorRow so it will trigger an `onChangeDataSource` when you change datasource instead of assigning it directly to the `query.datasource` property. This gives us the flexibility to reuse the same component in the alerting scenario where we set the `datasourceUid` property outside of the dataQuery object.

### CODE REVIEW

File: public/app/features/query/components/QueryEditorRows.tsx [hunk 1]
```diff
@@ -67,21 +85,29 @@ export class QueryEditorRows extends PureComponent<Props> {
...
+
+                  return (
+                    <QueryEditorRow
+                      id={query.refId}
+                      index={index}
+                      key={query.refId}
+                      data={data}
+                      query={query}
+                      dataSourceSettings={dataSourceSettings}
+                      onChangeDataSourceSettings={onChangeDataSourceSettings}
```
Comment (torkelo) [inline public/app/features/query/components/QueryEditorRows.tsx:102]: not sure I like the naming of these, or at lleast onChangeDataSourceSettings feels very misleading, it has nothing to do with changing data source settings. the main QueryEditorRows calls this dsSettings , and the change callback should be onChangeDataSource

File: public/app/features/query/components/QueryEditorRowTitle.tsx [hunk 2]
```diff
@@ -2,37 +2,36 @@ import React, { useState } from 'react';
...
-  queries,
-  onClick,
-  onChange,
-  onTimeRangeChange,
-  timeRange,
-  collapsedText,
-}) => {
-  const theme = useTheme();
-  const styles = getQueryEditorRowTitleStyles(theme);
+export const QueryEditorRowTitle: React.FC<Props> = (props) => {
```
Comment (torkelo) [inline public/app/features/query/components/QueryEditorRowTitle.tsx:22]: Maybe rename this to QueryEditorRowHeader? feels like it does more than the title :)

File: public/app/types/unified-alerting-dto.ts [hunk 3]
```diff
@@ -92,16 +92,12 @@ export enum GrafanaAlertState {
...
-  datasource: string;
-  datasourceUid: string;
-}
 export interface GrafanaQuery {
   refId: string;
   queryType: string;
   relativeTimeRange: RelativeTimeRange;
-  model: GrafanaQueryModel;
+  datasourceUid: string;
+  model: DataQuery;
```
Comment (peterholmberg) [inline public/app/types/unified-alerting-dto.ts:100]:
```suggestion
  model: DataQuery | ExpressionQuery;
```
?

File: public/app/features/query/components/QueryEditorRows.tsx [hunk 4]
```diff
@@ -67,21 +85,29 @@ export class QueryEditorRows extends PureComponent<Props> {
...
-                    onChange={(query) => this.onChangeQuery(query, index)}
-                    onRemoveQuery={this.onRemoveQuery}
-                    onAddQuery={this.props.onAddQuery}
-                    onRunQuery={this.props.onRunQueries}
-                    queries={queries}
-                  />
-                ))}
+                {queries.map((query, index) => {
+                  const dataSourceSettings = getDataSourceSettings(query, dsSettings);
+                  const onChangeDataSourceSettings = dsSettings.meta.mixed
```
Comment (mckn) [inline public/app/features/query/components/QueryEditorRows.tsx:90]: @torkelo currently I check if the callback is set to know if I should render the data source picker. I'm leaning towards adding another boolean flag to the component to control this to prevent writing code like the one I'm referencing here. What is your thoughts on this?

File: public/app/features/query/components/QueryEditorRows.tsx (hunk 1 above)
Comment (torkelo) [reply public/app/features/query/components/QueryEditorRows.tsx:102]: think that is ok (to check onDataSourceChange)

File: public/app/types/unified-alerting-dto.ts (hunk 3 above)
Comment (mckn) [reply public/app/types/unified-alerting-dto.ts:100]: I don't think we need to do this since the `ExpressionQuery` is a `DataQuery`.

Comment (peterholmberg) [reply public/app/types/unified-alerting-dto.ts:100]: but we won't have access to `ExpressionQuery` props when doing something like `model.expression` ?

File: public/app/features/alerting/unified/hooks/useFilteredRules.ts [hunk 5]
```diff
@@ -45,11 +47,7 @@ const reduceNamespaces = (filters: RuleFilterState) => {
 const reduceGroups = (filters: RuleFilterState) => {
   return (groupAcc: CombinedRuleGroup[], group: CombinedRuleGroup) => {
     const rules = group.rules.filter((rule) => {
-      if (
-        filters.dataSource &&
-        isGrafanaRulesSource(rule.namespace.rulesSource) &&
-        !rule.queries?.find(({ datasource }) => datasource === filters.dataSource)
-      ) {
+      if (filters.dataSource && isGrafanaRulerRule(rule.rulerRule) && !isQueryingDataSource(rule.rulerRule, filters)) {
```
Comment (domasx2) [inline public/app/features/alerting/unified/hooks/useFilteredRules.ts:50]: I would change it to `if (filters.dataSource && !(isGrafanaRulerRule(rule.rulerRule) && isQueryingDataSource(rule.rulerRule, filters))) {` 
to filter out rules both wher datasource doesn't match and where we don't have the grafana rule.
Though seems improbable this would ever be an issue

File: public/app/features/alerting/unified/hooks/useFilteredRules.ts [hunk 6]
```diff
@@ -84,3 +82,21 @@ const reduceGroups = (filters: RuleFilterState) => {
     return groupAcc;
   };
 };
+
+const isQueryingDataSource = (rulerRule: RulerGrafanaRuleDTO, filter: RuleFilterState): boolean => {
+  if (!filter.dataSource) {
+    return true;
+  }
+
+  if (!Array.isArray(rulerRule.grafana_alert.data)) {
```
Comment (domasx2) [inline public/app/features/alerting/unified/hooks/useFilteredRules.ts:91]: Imho it's not worthwhile to pepper code guarding against `null` that is not normally expected and does not match types. It's boilerplaty, have to do it in a lot of places, hides the problem rather than fixes it.

I suggest to either:
* Just let it fail to become aware of the problem faster, and fix the backend issue
* fill in the possible null at the api endpoint level, in `api/ruler.ts`

**Review by domasx2 — APPROVED** [review]
some minor quibbles, but if it works, it works :+1:

**Comment by mckn** [comment]
> some minor quibbles, but if it works, it works 👍

Great feedback, will have a look!

File: public/app/types/unified-alerting-dto.ts (hunk 3 above)
Comment (mckn) [reply public/app/types/unified-alerting-dto.ts:100]: I think we have to do a guard check to know what type of model we have. But that needs to be done regardless of the `model: DataQuery` or `model: DataQuery | ExpressionQuery` if I'm not thinking totally wrong.

File: public/app/features/alerting/unified/hooks/useFilteredRules.ts (hunk 5 above)
Comment (mckn) [reply public/app/features/alerting/unified/hooks/useFilteredRules.ts:50]: So I am basically translating that into:

- If we have a filter.dataSource and the rule is an alerting rule then is should be filtered out.
- If we have a filter.dataSource and the rule is an grafana rule then check if the rule is querying the dataSource.

Is that correct?

Comment (domasx2) [reply public/app/features/alerting/unified/hooks/useFilteredRules.ts:50]: hmm wait no, your version was correct. This would fitler out cloud alerts as well. My bad :)

Comment (mckn) [reply public/app/features/alerting/unified/hooks/useFilteredRules.ts:50]: hmm, I think I need to change it a bit. Right?

Comment (mckn) [reply public/app/features/alerting/unified/hooks/useFilteredRules.ts:50]: So we are basically filtering out rules in two different places here witch feels a bit hard to follow.

Comment (domasx2) [reply public/app/features/alerting/unified/hooks/useFilteredRules.ts:50]: What I wanted to do is defien the case where it's a grafana rule, but only has `promRule` and not `rulerRule`, to filter it out in such case. But that's an edge case and *should never happen* (tm) 

But we **should not** filter out cloud alerting rules here. Datasource filtering for those is done at namespace level. This is because  cloud rules in a given namespace all belong to the same datasource, there is no datasource property on the rule itself.  S oon rule level datasource should not be conisdered  for cloud rules and they should pass.  

So I think the existing code is good enough

Comment (mckn) [reply public/app/features/alerting/unified/hooks/useFilteredRules.ts:50]: I will leave it as it is and we probably can clean this code up a little bit when we have everything else in place.

Comment (mckn) [reply public/app/features/alerting/unified/hooks/useFilteredRules.ts:50]: Sounds ok?

Comment (domasx2) [reply public/app/features/alerting/unified/hooks/useFilteredRules.ts:50]: yup :+1:  sorry for muddying the waters here
